### PR TITLE
Router settle with balance

### PR DESCRIPTION
**Pull request by hensha256**
Allow settle to be funds from the router

### CODE REVIEW

File: test/shared/implementation/V4RouterImplementation.sol [hunk 1]
```diff
@@ -27,7 +27,11 @@ contract V4RouterImplementation is V4Router, ReentrancyLock {
     }
 
     function _pay(Currency token, address payer, uint256 amount) internal override {
-        ERC20(Currency.unwrap(token)).safeTransferFrom(payer, address(poolManager), amount);
+        if (payer == address(this)) {
+            ERC20(Currency.unwrap(token)).safeTransfer(address(poolManager), amount);
```
[the remark this answers is not in the document]
Comment (hensha256) [reply test/shared/implementation/V4RouterImplementation.sol:31]: 🤷‍♀️ its a test so didnt think about it lol, will do it now

[the remark this answers is not in the document]
Comment (hensha256) [reply test/shared/implementation/V4RouterImplementation.sol]: done! i also moved this contract to `/mocks` and deleted `/implementations`

File: test/mocks/MockV4Router.sol [hunk 2]
```diff
@@ -27,7 +28,11 @@ contract V4RouterImplementation is V4Router, ReentrancyLock {
     }
 
     function _pay(Currency token, address payer, uint256 amount) internal override {
-        ERC20(Currency.unwrap(token)).safeTransferFrom(payer, address(poolManager), amount);
+        if (payer == address(this)) {
+            token.transfer(address(poolManager), amount);
```
Comment (snreynolds) [inline test/mocks/MockV4Router.sol:32]: so I think we both have the same problem in pay where by this point (if its been called from DeltaResolver) it definitely wont be NATIVE currency, but we still check if its native inside currency. I wonder if we should write a currency.transferERC20() without that native check.. anyways food for thought.

File: src/V4Router.sol [hunk 3]
```diff
@@ -44,11 +44,11 @@ abstract contract V4Router is IV4Router, BaseActionsRouter, DeltaResolver {
...
             if (action == Actions.SETTLE_ALL) {
                 Currency currency = params.decodeCurrency();
-                uint256 amount = _getFullSettleAmount(currency);
-
-                // TODO support address(this) paying too
                 // TODO should it have a maxAmountOut added slippage protection?
-                _settle(currency, _msgSender(), amount);
+                _settle(currency, _msgSender(), _getFullSettleAmount(currency));
+            } else if (action == Actions.SETTLE_WITH_BALANCE) {
+                Currency currency = params.decodeCurrency();
```
Comment (snreynolds) [inline src/V4Router.sol:50]: i like how you do this here, Im going to change posm to do this as well

**Review by snreynolds — APPROVED** [review]

File: test/mocks/MockV4Router.sol (hunk 2 above)
Comment (hensha256) [reply test/mocks/MockV4Router.sol:32]: this contract is actually just a mock for tests - in the universal router (where i implement `_pay`) it doesnt check if its ETH. But its a good thought for posm

**Review by snreynolds — APPROVED** [review]

**Pull request merged**